### PR TITLE
#167484587 user should be able to signout

### PR DESCRIPTION
**Pull request by raymond42**
#### What does this PR do?
 * User should be able to sign out

 #### Description of Task to be completed?
 
#### How should this be manually tested?
 * `git checkout -b ft-user-signout-#167484587 `
* `git pull origin ft-user-signout-#167484587 `
 * `npm install`
 * `npm run dev`
 * Using postman test using the following endpoints

1. /api/v1/users/signout -- include a logged in user token to be signed out
2. Test any other protected endpoint to check if user is signed out

#### What are the relevant pivotal tracker stories?
 [#167484587](https://www.pivotaltracker.com/story/show/167484587)

 #### Screenshots
 
Screen Shot 2019-08-15 at 16 30 06
Screen Shot 2019-08-15 at 16 30 56

### CODE REVIEW

**Review by malaba6 — CHANGES_REQUESTED** [review]
You did not add this endpoint in the swagger documentation, can u plz document it?

File: src/routes/routes.js [hunk 1]
```diff
@@ -19,5 +19,6 @@ router.get('/api/v1/verify-email/:token', verifyToken, UserAuth.verifyEmail);
 router.post('/api/v1/send-email', Validation.emailValidation, UserAuth.emailSender);
 router.post('/api/v1/reset-password/:token', verifyToken, Validation.passwordValidation, UserAuth.resetPassword);
 router.post('/api/v1/login', Validation.loginValidation, UserAuth.login);
+router.post('/api/v1/signout', [verifyToken], UserAuth.signoutUser);
```
Comment (malaba6) [inline src/routes/routes.js]: I think the route here should be `/api/v1/users/signout.

**Review by placideirandora — CHANGES_REQUESTED** [review]
1. I think there is no need to hash the identifier because when you store them in the database, one identifier is visible and the other one is encrypted. Therefore, If you want to encrypt it, you can store one encrypted identifier and you will be comparing their hashes in **verify token** middleware or you can store it unencrypted but only once because the hash is not playing any role.
Screen Shot 2019-08-16 at 10 57 15
Screen Shot 2019-08-16 at 10 57 29
Screen Shot 2019-08-16 at 11 03 03

**Comment by raymond42** [comment]
> 1. I think there is no need to hash the identifier because when you store them in the database, one identifier is visible and the other one is encrypted. Therefore, If you want to encrypt it, you can store one encrypted identifier and you will be comparing their hashes in **verify token** middleware or you can store it unencrypted but only once because the hash is not playing any role.
> 
> Screen Shot 2019-08-16 at 10 57 15
> 
> Screen Shot 2019-08-16 at 10 57 29
> 
> Screen Shot 2019-08-16 at 11 03 03

Thanks placide. However, I'm not hashing the identifier. I'm hashing the token and then extract the hashed token remove those numbers and store them as identifier because those numbers they never change but generally every time we hash a token it changes even if It's the same token but those numbers they never change. So that is why when I'm validating I take the token and hash it again and extract it and compare the identifiers if they are the same, that is when the system knows that this token is blacklisted land the person is logged out.

**Comment by placideirandora** [comment]
> > 1. I think there is no need to hash the identifier because when you store them in the database, one identifier is visible and the other one is encrypted. Therefore, If you want to encrypt it, you can store one encrypted identifier and you will be comparing their hashes in **verify token** middleware or you can store it unencrypted but only once because the hash is not playing any role.
> > 
> > Screen Shot 2019-08-16 at 10 57 15
> > Screen Shot 2019-08-16 at 10 57 29
> > Screen Shot 2019-08-16 at 11 03 03
> 
> Thanks placide. However, I'm not hashing the identifier. I'm hashing the token and then extract the hashed token remove those numbers and store them as identifier because those numbers they never change but generally every time we hash a token it changes even if It's the same token but those numbers they never change. So that is why when I'm validating I take the token and hash it again and extract it and compare the identifiers if they are the same, that is when the system knows that this token is blacklisted land the person is logged out.

I think the token hash
 is not playing any role because I have tested that I can log the user out and completely destroy the token by using only the identifier. The token hash is stored for nothing because you don't need it when you want to verify if the token is valid or destroyed. You only use the identifier!
Screen Shot 2019-08-16 at 12 37 06
Screen Shot 2019-08-16 at 12 30 16
Screen Shot 2019-08-16 at 12 30 58
Screen Shot 2019-08-16 at 12 31 11
Screen Shot 2019-08-16 at 12 30 48

File: src/routes/routes.js (hunk 1 above)
Comment (raymond42) [reply src/routes/routes.js]: thanks malaba. done implementing it

**Comment by raymond42** [comment]
> > > 1. I think there is no need to hash the identifier because when you store them in the database, one identifier is visible and the other one is encrypted. Therefore, If you want to encrypt it, you can store one encrypted identifier and you will be comparing their hashes in **verify token** middleware or you can store it unencrypted but only once because the hash is not playing any role.
> > > 
> > > Screen Shot 2019-08-16 at 10 57 15
> > > Screen Shot 2019-08-16 at 10 57 29
> > > Screen Shot 2019-08-16 at 11 03 03
> > 
> > 
> > Thanks placide. However, I'm not hashing the identifier. I'm hashing the token and then extract the hashed token remove those numbers and store them as identifier because those numbers they never change but generally every time we hash a token it changes even if It's the same token but those numbers they never change. So that is why when I'm validating I take the token and hash it again and extract it and compare the identifiers if they are the same, that is when the system knows that this token is blacklisted land the person is logged out.
> 
> I think the token hash
> is not playing any role because I have tested that I can log the user out and completely destroy the token by using only the identifier. The token hash is stored for nothing because you don't need it when you want to verify if the token is valid or destroyed. You only use the identifier!
> Screen Shot 2019-08-16 at 12 37 06
> Screen Shot 2019-08-16 at 12 30 16
> Screen Shot 2019-08-16 at 12 30 58
> Screen Shot 2019-08-16 at 12 31 11
> Screen Shot 2019-08-16 at 12 30 48

1. From the research I made, It's not good practice to store token as it is, you have to hash it for security purpose.
2. I'm getting the identifier from the token. there on your screenshot, you are saving the identifier in the database only but where you are getting that identifier is from the hashed token, so you can't get that identifier when you don't have a hashed  token

**Comment by placideirandora** [comment]
> > > > 1. I think there is no need to hash the identifier because when you store them in the database, one identifier is visible and the other one is encrypted. Therefore, If you want to encrypt it, you can store one encrypted identifier and you will be comparing their hashes in **verify token** middleware or you can store it unencrypted but only once because the hash is not playing any role.
> > > > 
> > > > Screen Shot 2019-08-16 at 10 57 15
> > > > Screen Shot 2019-08-16 at 10 57 29
> > > > Screen Shot 2019-08-16 at 11 03 03
> > > 
> > > 
> > > Thanks placide. However, I'm not hashing the identifier. I'm hashing the token and then extract the hashed token remove those numbers and store them as identifier because those numbers they never change but generally every time we hash a token it changes even if It's the same token but those numbers they never change. So that is why when I'm validating I take the token and hash it again and extract it and compare the identifiers if they are the same, that is when the system knows that this token is blacklisted land the person is logged out.
> > 
> > 
> > I think the token hash
> > is not playing any role because I have tested that I can log the user out and completely destroy the token by using only the identifier. The token hash is stored for nothing because you don't need it when you want to verify if the token is valid or destroyed. You only use the identifier!
> > Screen Shot 2019-08-16 at 12 37 06
> > Screen Shot 2019-08-16 at 12 30 16
> > Screen Shot 2019-08-16 at 12 30 58
> > Screen Shot 2019-08-16 at 12 31 11
> > Screen Shot 2019-08-16 at 12 30 48
> 
> 1. From the research I made, It's not good practice to store token as it is, you have to hash it for security purpose.
> 2. I'm getting the identifier from the hashed token. there on your screenshot, you are saving the identifier in the database only but where you are getting that identifier is from the hashed token, so you can't get that identifier when you don't have a hashed  token

No. You're getting the identifier from the request not the hashed token, **req.headers.authorization**!

**Comment by raymond42** [comment]
> > > > > 1. I think there is no need to hash the identifier because when you store them in the database, one identifier is visible and the other one is encrypted. Therefore, If you want to encrypt it, you can store one encrypted identifier and you will be comparing their hashes in **verify token** middleware or you can store it unencrypted but only once because the hash is not playing any role.
> > > > > 
> > > > > Screen Shot 2019-08-16 at 10 57 15
> > > > > Screen Shot 2019-08-16 at 10 57 29
> > > > > Screen Shot 2019-08-16 at 11 03 03
> > > > 
> > > > 
> > > > Thanks placide. However, I'm not hashing the identifier. I'm hashing the token and then extract the hashed token remove those numbers and store them as identifier because those numbers they never change but generally every time we hash a token it changes even if It's the same token but those numbers they never change. So that is why when I'm validating I take the token and hash it again and extract it and compare the identifiers if they are the same, that is when the system knows that this token is blacklisted land the person is logged out.
> > > 
> > > 
> > > I think the token hash
> > > is not playing any role because I have tested that I can log the user out and completely destroy the token by using only the identifier. The token hash is stored for nothing because you don't need it when you want to verify if the token is valid or destroyed. You only use the identifier!
> > > Screen Shot 2019-08-16 at 12 37 06
> > > Screen Shot 2019-08-16 at 12 30 16
> > > Screen Shot 2019-08-16 at 12 30 58
> > > Screen Shot 2019-08-16 at 12 31 11
> > > Screen Shot 2019-08-16 at 12 30 48
> > 
> > 
> > 
> > 1. From the research I made, It's not good practice to store token as it is, you have to hash it for security purpose.
> > 2. I'm getting the identifier from the hashed token. there on your screenshot, you are saving the identifier in the database only but where you are getting that identifier is from the hashed token, so you can't get that identifier when you don't have a hashed  token
> 
> No. You're getting the identifier from the request not the hashed token, **req.headers.authorization**!

Ooooh I get now what you mean

**Comment by placideirandora** [comment]
> > > > > > 1. I think there is no need to hash the identifier because when you store them in the database, one identifier is visible and the other one is encrypted. Therefore, If you want to encrypt it, you can store one encrypted identifier and you will be comparing their hashes in **verify token** middleware or you can store it unencrypted but only once because the hash is not playing any role.
> > > > > > 
> > > > > > Screen Shot 2019-08-16 at 10 57 15
> > > > > > Screen Shot 2019-08-16 at 10 57 29
> > > > > > Screen Shot 2019-08-16 at 11 03 03
> > > > > 
> > > > > 
> > > > > Thanks placide. However, I'm not hashing the identifier. I'm hashing the token and then extract the hashed token remove those numbers and store them as identifier because those numbers they never change but generally every time we hash a token it changes even if It's the same token but those numbers they never change. So that is why when I'm validating I take the token and hash it again and extract it and compare the identifiers if they are the same, that is when the system knows that this token is blacklisted land the person is logged out.
> > > > 
> > > > 
> > > > I think the token hash
> > > > is not playing any role because I have tested that I can log the user out and completely destroy the token by using only the identifier. The token hash is stored for nothing because you don't need it when you want to verify if the token is valid or destroyed. You only use the identifier!
> > > > Screen Shot 2019-08-16 at 12 37 06
> > > > Screen Shot 2019-08-16 at 12 30 16
> > > > Screen Shot 2019-08-16 at 12 30 58
> > > > Screen Shot 2019-08-16 at 12 31 11
> > > > Screen Shot 2019-08-16 at 12 30 48
> > > 
> > > 
> > > 
> > > 1. From the research I made, It's not good practice to store token as it is, you have to hash it for security purpose.
> > > 2. I'm getting the identifier from the hashed token. there on your screenshot, you are saving the identifier in the database only but where you are getting that identifier is from the hashed token, so you can't get that identifier when you don't have a hashed  token
> > 
> > 
> > No. You're getting the identifier from the request not the hashed token, **req.headers.authorization**!
> 
> Ooooh I get you now what you mean

You're Welcome!

**Review by placideirandora — APPROVED** [review]

File: src/controllers/userController.js [hunk 2]
```diff
@@ -138,6 +138,24 @@ class Authentication {
...
+   * @param {object} res
+   * @returns {object} logged out user
+   */
+  static async signoutUser(req, res) {
+    const token = req.headers.authorization;
+    const identifier = token.match(/\d+/g).join(''); // Extract numbers only from token to be used to uniquely identify a token in db
+    const rejectedToken = await DroppedToken.create({ identifier });
+
+    return res.status(200).json({
+      status: 200,
```
Comment (mkiterian) [inline src/controllers/userController.js]: remove this field

File: src/controllers/userController.js [hunk 3]
```diff
@@ -138,6 +138,24 @@ class Authentication {
...
+      message: 'Successfully logged out.',
+      data: rejectedToken
```
Comment (mkiterian) [inline src/controllers/userController.js]: why send back this data

File: src/controllers/userController.js [hunk 4]
```diff
@@ -138,6 +138,24 @@ class Authentication {
...
     })(req, res);
   }
+
+  /**
+   * @description user logout
+   * @param {object} req
+   * @param {object} res
+   * @returns {object} logged out user
+   */
+  static async signoutUser(req, res) {
```
Comment (mkiterian) [inline src/controllers/userController.js]: for consistency use logout since to get into the app we are going with login

**Review by mkiterian — CHANGES_REQUESTED** [review]

File: src/controllers/userController.js (hunk 2 above)
Comment (raymond42) [reply src/controllers/userController.js]: done

File: src/controllers/userController.js (hunk 3 above)
Comment (raymond42) [reply src/controllers/userController.js]: I've removed it. thanks

File: src/controllers/userController.js (hunk 4 above)
Comment (raymond42) [reply src/controllers/userController.js]: done

**Pull request merged**